pkg/cueloader/announcer.go

```diff
@@ -0,0 +1,125 @@
+package cueloader
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+
+	"github.com/elnormous/contenttype"
+)
+
+// Announcer is a simple http.Handler that announces the latest value for a []byte via text/event-stream. In scenarios
+// where there are many new []byte values to announce, it prioritizes sending the latest value byte rather than every
+// update.
+type Announcer struct {
+	b  []byte
+	mu *sync.Mutex
+	contentType string
+	chs map[chan struct{}]struct{}
+}
+
+func NewAnnouncer(contentType string) *Announcer {
+	return &Announcer{
+		chs: map[chan struct{}]struct{}{},
+		mu:  &sync.Mutex{},
+		contentType: contentType,
+	}
+}
+
+func (a *Announcer) listen(l chan struct{}) ([]byte, func()) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.chs[l] = struct{}{}
+
+	return a.b, func() {
+		a.mu.Lock()
+		defer a.mu.Unlock()
+		delete(a.chs, l)
+	}
+}
+
+func (a *Announcer) get() []byte {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	return a.b
+}
+
+func (a *Announcer) Announce(b []byte) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.b = b
+
+	for l := range a.chs {
+		select {
+		case l <- struct{}{}:
+		default:
+		}
+	}
+}
```

Comment:
Hmm, I think there's potential for some unexpected behavior like duplicated output here, e.g.:
```
Announce("a") // sends on channel
<-ch // unblocks
Announce("b") // puts another struct into the channel buffer
m.get() // returns "b"
<-ch // unblocked based on pending item in buffer
m.get() // returns "b"
```

Reply:
Yes. In this case we care most about:
- Avoiding a slow/hung TCP connection from blocking everyone else
- Never dropping the most recent event because a channel was full and failing to notify a listener of an event

At the cost of a bit extra memory, we could check if b is different from the last value we wrote. At the cost of a slightly more confusing interface (but not more memory), we could return two values from `get()`, the `[]byte` and an opaque `*struct{}` version token. If the version token matches, it's a noop. If the version token changes, we re-announce.

What do you think?

Reply:
I posted an alternative in #133 that I think solves the blocking issue without duplicating messages.

Reply:
I looked at #133. It's both more complicated to understand and based on my reading doesn't actually solve the problem. In a rapid series of edits it will send an old message before sending a new one. Perhaps I'm reading it wrong though.